lib/collection/src/shards/forward_proxy_shard.rs

```diff
@@ -196,6 +168,146 @@ impl ForwardProxyShard {
+    /// next batch from.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe.
+    async fn read_batch(
+        &self,
+        offset: Option<PointIdType>,
+        batch_size: usize,
+        runtime_handle: &Handle,
+    ) -> CollectionResult<(Vec<PointStructPersisted>, Option<PointIdType>)> {
+        let limit = batch_size + 1;
+
+        let mut batch = self
+            .wrapped_shard
+            .scroll_by(
+                offset,
+                limit,
+                &WithPayloadInterface::Bool(true),
+                &WithVector::Bool(true),
+                None,
+                runtime_handle,
+                None,
+                None,                           // No timeout
+                HwMeasurementAcc::disposable(), // Internal operation, no need to measure hardware here.
+            )
+            .await?;
+
+        let next_page_offset = (batch.len() >= limit).then(|| batch.pop().unwrap().id);
+
+        let points = batch
+            .into_iter()
+            .map(PointStructPersisted::try_from)
+            .collect::<Result<Vec<PointStructPersisted>, String>>()?;
+
+        Ok((points, next_page_offset))
+    }
+
+    /// Read a batch of points using a hash ring to transfer to the remote shard
+    ///
+    /// Only the points that satisfy the hash ring filter will be transferred.
+    ///
+    /// This applies oversampling in case of resharding to account for points that will be filtered
+    /// out by the hash ring. Each batch of points should therefore be roughly `batch_size`, but it
+    /// may be a bit smaller or larger.
+    ///
+    /// It is optimized for reading and transferring only a fraction of the points in this shard by
+    /// using a hash ring. If you need to read and transfer 100% of the points, use [`read_batch`]
+    /// instead.
+    ///
+    /// Returns batch of points and new point offset. The new point offset can be used to start the
+    /// next batch from.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe.
+    async fn read_batch_with_hashring(
+        &self,
+        offset: Option<PointIdType>,
+        batch_size: usize,
+        hashring_filter: &HashRingRouter,
+        runtime_handle: &Handle,
+    ) -> CollectionResult<(Vec<PointStructPersisted>, Option<PointIdType>)> {
+        // Oversample batch size to account for points that will be filtered out by the hash ring
+        let oversample_factor = match &hashring_filter {
+            HashRingRouter::Single(_) => 1,
+            // - resharding: 1 -> 2, transfer 50%,  factor 2
+            // - resharding: 2 -> 3, transfer 33%,  factor 3
+            // - resharding: 3 -> 4, transfer 25%,  factor 4
+            // - resharding: 2 -> 1, transfer 100%, factor 1
+            // - resharding: 3 -> 2, transfer 50%,  factor 2
+            // - resharding: 4 -> 3, transfer 33%,  factor 3
+            HashRingRouter::Resharding { old: _, new } => new.len().max(1),
```

Comment:
```suggestion
            HashRingRouter::Resharding { old: _, new } => new.len(),
```

What's the use of `.max(1)` here?  `new.len()` will always >= 1. This will limit the factor to 1 and defeat the purpose of oversampling.

Reply:
Okay nvm.`max(1)` will ensure that the final value remains >= 1. What I described would be the behaviour of `.min(1)`.

Reply:
:+1: Yeah, I'm also always confused by min/max in practice.

Even though we should never reach 0 shards, it's just there to stay on the safe side.

Reply:
I always use `cmp::max`/`cmp::min` explicitly, cause `.max(...)` confuses me. :/

```suggestion
            HashRingRouter::Resharding { old: _, new } => cmp::max(1, new.len()),
```